File: java_solutions/2435.paths-in-matrix-whose-sum-is-divisible-by-k.java
/*
 * @lc app=leetcode id=2435 lang=java
 *
 * [2435] Paths in Matrix Whose Sum Is Divisible by K
 */

class Solution { public int numberOfPaths(int[][] grid, int k) { return ((Function<int[][][], Integer>) dp -> IntStream.range(0, grid.length).map(i -> IntStream.range(0, grid[0].length).map(j -> (i == 0 && j == 0) ? (dp[0][0][grid[0][0] % k] = 1) * 0 : IntStream.range(0, k).map(r -> dp[i % 2][j][(r + grid[i][j]) % k] = ((i > 0 ? dp[1 - i % 2][j][r] : 0) + (j > 0 ? dp[i % 2][j - 1][r] : 0)) % 1000000007).sum() * 0).sum() * 0).sum() * 0 == 0 ? dp[(grid.length - 1) % 2][grid[0].length - 1][0] : 0).apply(new int[2][grid[0].length][k]); } }

File: kotlin_solutions_1/2435.paths-in-matrix-whose-sum-is-divisible-by-k.kt
/*
 * @lc app=leetcode id=2435 lang=kotlin
 *
 * [2435] Paths in Matrix Whose Sum Is Divisible by K
 */
class Solution { fun numberOfPaths(grid: Array<IntArray>, k: Int): Int = Array(grid[0].size) { IntArray(k) }.let { dp -> grid.indices.forEach { i -> IntArray(grid[0].size * k).let { nextRow -> grid[0].indices.forEach { j -> (0 until k).forEach { r -> nextRow[j * k + (r + grid[i][j]) % k] = ((if (i == 0 && j == 0) (if (r == 0) 1 else 0) else (if (i > 0) dp[j][r] else 0) + (if (j > 0) nextRow[(j - 1) * k + r] else 0)) % 1000000007) } } }.run { (0 until grid[0].size).forEach { j -> (0 until k).forEach { r -> dp[j][r] = it[j * k + r] } } } }.run { dp[grid[0].size - 1][0] } }